### ISSUE TITLE
Allow WAN Interface to be enabled/disabled based on CARP Status

### ISSUE THREAD

**Issue by shehzman** [opnsense/core#7333]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md

**Is your feature request related to a problem? Please describe.**

Currently, if a user wants to setup a passive/active router setup (high availability) on the WAN interface, they will need 3 static IP addresses on the WAN side so they can setup CARP. Due to the shortage of IPV4 addresses and many ISP's not even distributing an IPV6 addresses yet, it may be very expensive or outright impossible for some users to obtain 3 static WAN IP addresses. Especially in residential environments, such as a homelab, where ISP's typically distribute only a single dynamic IP address per line. 

**Describe the solution you like**

Add an option to enable/disable the WAN interface based on the CARP status of a selected interface. Very similar to the Depend On (CARP) option in Wireguard instances and this [custom script](https://gist.github.com/spali/2da4f23e488219504b2ada12ac59a7dc). This way, users can have single and or dynamic IP addresses and still achieve a high availability OPNSense setup. On the interfaces/WAN page, maybe add a dropdown that allows the user to select which CARP interface the WAN will depend on. If the CARP status is master, the WAN is enabled. If it is backup, WAN will be disabled. Also, if applicable, a way for the router currently in backup mode to access the internet. Maybe similar to how the custom script is able to accomplish this. 

**Describe alternatives you considered**
To achieve high availability on the WAN side, a user can add another router in front of the OPNSense instances and configure the DMZ settings accordingly. However, I would say this adds too much complexity and defeats the purpose of a high availability router setup. 

**Additional context**

N/A

**Comment by skl283** [comment]
Just cameto this topic - I am currently in the process of implementing exactly this with the custom script you mentioned - it would be very nice if opn has the Option to enable/disable the WAN interface based on the CARP status of a selected interface.

Since there is for the PPOE Connection Type already an trigger to connect/disconnect connections i assume this should be absolutely in scope to extend the functionality to disable an interface or to select more than one Interface to disable more than just one Interface.

<img width="1800" alt="image" src="https://github.com/opnsense/core/assets/53855219/2bfb56b4-7fe6-4067-830c-d5996b32d992">

@shehzman i'll try to use the following edited script with more logging to archive this: https://gist.github.com/willjasen/6ae0f47bca36ced2bd52b2fefc2bc21e

**Comment by oasis9** [comment]
I'm very interested in this use case. I don't need instant/seamless failover but my ISP forces me to use DHCP even though I have a static reservation. I need OPNsense to disable WAN on the backup and only enable it when the backup enters CARP master state. Someone wrote a script but it keeps getting broken by updates. Would be a lot nicer if this was natively supported.

**Comment by willjasen** [comment]
[The script I have](https://gist.github.com/willjasen/6ae0f47bca36ced2bd52b2fefc2bc21e) seems to be working for me on the latest versions of OPNsense, but I am a sample of 1. In the way that it operates, it will only enable WAN interfaces once all CARP subsystems are in the MASTER state and it will disable all WAN interfaces if even one CARP subsystem is not MASTER. For some further clarification, the colloquial use case is for WAN interfaces but the script makes no distinction other than an array of interfaces to be toggled as described above.

**Comment by oasis9** [comment]
@willjasen I'm a bit busy today but when I get time I'll test your solution, looks like it's just what I need. Will report back if it works for me 😊

**Comment by skl283** [comment]
@AdSchellevis @fichtner could you please look at this Feature Req.?

Please look also at my Post: https://github.com/opnsense/core/issues/7333#issuecomment-2027884872 i think that this req. is fully at OPNsense‘s scope and perhaps not too difficult to implement?

**Comment by fichtner** [comment]
I think modifying config.xml state on CARP state changes is neither desirable nor wise.


Cheers,
Franco

**Comment by fichtner** [comment]
And before I forget...

> they will need 3 static IP addresses on the WAN side so they can setup CARP

You can get by with just one CARP IP alias and IPv4 set to "none".

**Comment by shehzman** [comment]
> I think modifying config.xml state on CARP state changes is neither desirable nor wise.
> 
> Cheers, Franco

Is this not how you modify the WireGuard interface when the CARP state changes if the user uses the Depend on (CARP) option? That’s pretty much what I’m requesting but for the WAN interface. This is coming from a place of ignorance as I know next to nothing about the inner workings of OPNSense’s code base so please correct me if I’m wrong.

**Comment by shehzman** [comment]
> And before I forget...
> 
> > they will need 3 static IP addresses on the WAN side so they can setup CARP
> 
> You can get by with just one CARP IP alias and IPv4 set to "none".

Wouldn’t the single CARP IP alias need to be static?  Otherwise it sounds like you’d have to change it every time a dynamic WAN IP changes.

**Comment by fichtner** [comment]
Some community wireguard carp scripts have done this perhaps, but it's not good to modify configuration state and the core code supporting this does it differently for reasons mentioned.

CARP is always static?!


Cheers,
Franco

**Comment by shehzman** [comment]
> Some community wireguard carp scripts have done this perhaps, but it's not good to modify configuration state and the core code supporting this does it differently for reasons mentioned.
> 
> CARP is always static?!
> 
> Cheers, Franco

Sorry I’m a little confused. The Depend On (CARP) feature I’m referring to is an official feature in OPNsense for the WireGuard interface and not a community script. One of my OPNsense setups has a proper CARP setup with this feature. If I power off of my primary OPNsense instance or manually set the CARP status to backup, the WireGuard interface on the primary instance will be disabled and the WireGuard interface on the secondary OPNsense instance will be enabled automatically. WireGuard will then be disabled on the secondary instance and enabled on the primary instance when it has a CARP status of master. Is there something specific about the WAN interface that wouldn’t allow an existing feature like the one I just highlighted to work there too? 

From my understanding, CARP/VRRP in general does not support dynamic IP addresses.

**Comment by fichtner** [comment]
_Stopped_, not _disabled_. Runtime condition (CARP state) causes runtime modification of service (stop and start accordingly).

**Comment by shehzman** [comment]
> _Stopped_, not _disabled_. Runtime condition (CARP state) causes runtime modification of service (stop and start accordingly).
> 
> To illustrate what I mean from the OP's script:
> 
> https://gist.github.com/spali/2da4f23e488219504b2ada12ac59a7dc#file-10-wancarp-L25-L26
> 
> Runtime value causes configuration change causes a reload wich could cause runtime value change which could cause configuration change which ...

Ahh I see. Thanks for the clarification, looks like the proper way to have a high availability WAN interface is to have 3 static WAN IP addresses or just manually enabling/disabling the WAN interface on each of your OPNsense systems. You could also have a single CARP alias and set the WAN IPv4 to none like you mentioned, but I think you need to change the CARP alias IP every time your WAN IP changes. Unless it’s possible to write a script to manually do that.

**Comment by fichtner** [comment]
If you use CARP on WAN you will definitely have a static address even if hidden behind a NAT somewhere... I'm not sure why this confuses you the way it does.

**Comment by shehzman** [comment]
> If you use CARP on WAN you will definitely have a static address even if hidden behind a NAT somewhere... I'm not sure why this confuses you the way it does.

Maybe it’s just a gap in my networking knowledge as I’m very much still a beginner at all this (mainly a dev). From what I know about CARP/VRRP, it requires at least 3 static IP addresses. One for each system and one for the shared IP address. On the WAN side, if you don’t have 3 static IP addresses, the only workaround I know and have seen other people use is to put a router (typically the one provided by the ISP) in front of your two OPNsense instances and configure the DMZ on it to point to the shared IP address. Not the biggest fan of this approach since it requires an extra router and can add extra complexity to a setup with a double NAT (even though port forwarding is taken care of with the DMZ). 

Maybe there’s another workaround I’m not familiar with or the router behind OPNsense solution isn’t as bad as I think it is, but that’s where I’m coming from in terms of my current knowledge.

**Comment by fichtner** [comment]
Ok no problem. For most purposes you don't need CARP on WAN really. This is mainly a workaround for suboptimal WAN access (like one DHCP account only) or when you want traffic to flow in on a HA fashion (but there are better technologies for it like OpenVPN multi home). In the bulk of use cases you want HA for your internal devices only. This is the main "router" case. The other one includes "server" side use and associated complications (that e.g. reverse proxies try solve).

**Comment by fichtner** [comment]
Don't you have some sort of VPN to terminate on one of the boxes? Then you really don't need the CARP on the WAN side.

**Comment by CoMPaTech** [comment]
The gist which was referred somewhere is something I use(d) to some extend (something else fails so not actively using it unfortunately). Where the set-up is to indeed have a single DHCP option on the WAN side, but requiring HA on hardware level. The full disclosure there is; two DHCP WAN providers for optimal breakout and CARP for deciding the primary OPNsense box to DHCPclient and get the WAN addresses.
That way if either OPNsense has a hardware issue or any of the WAN providers has an issue, all is still well. (Using routing metrics and gateway monitoring etc. for the WAN dual homing).
This requires the interface change though since one of these providers fixes by mac address, not client identifier.

**Comment by oasis9** [comment]
My ISP gives me a single reserved IPv4 address and an IPv6 prefix but neither of these work without renewing DHCP lease every 30 minutes, otherwise I get booted. So I can't use CARP on WAN but why should I not be able to benefit from failover of some sort? I don't care if my connections stop for a few seconds, a few seconds is better than having to manually enable WAN on my backup firewall. If modifying config.xml isn't the right way to do this, what is? I really like the loop of @willjasen's script, though I've made a comment on their gist about a call to interface_configure after unsetting the enable key in the interface's config, since this seems to prevent interface_configure from doing anything meaningful. I've got time to test this out today so I'll try their script out without those invocations and see how I go. Was wondering if interface_reset might be necessary but I'll admit I have next to no idea what best practice is here. Would be nice if this functionality was built in like CARP is.

**Comment by skl283** [comment]
thank you @fichtner for your comments on this Feature Request.

I also want to have an Solution for having an redundancy in Hard- & Software with 2 opn Machines like @CoMPaTech and @oasis9 and the others have already mentioned. For dialup Devices this is already implemented - like i posted above:

<img alt="image" width="1800" src="https://private-user-images.githubusercontent.com/53855219/318160959-2bfb56b4-7fe6-4067-830c-d5996b32d992.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM2MDI5NzQsIm5iZiI6MTcxMzYwMjY3NCwicGF0aCI6Ii81Mzg1NTIxOS8zMTgxNjA5NTktMmJmYjU2YjQtN2ZlNi00MDY3LTgzMGMtZDU5OTZiMzJkOTkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIwVDA4NDQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwZjUzZjI1ODMzODAxY2QxNmEzY2YzYmE4NmYxZWU4OTdlMzFkY2JiZDI3YzE0ZGNmMmI5NWIxNmM4OTFhNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tQggbDDyUh1UN4f1wqhI-htG15eVdBfcytfkaWxSbuQ">

it "only" lacks of doing DHCP Request on WAN sides only at **1** (the Master) opn-Machine! This should be absolutely in scope to extend the existing functionality to doing only at the Master Box DHCP Request, disable an interface or to select more than one Interface to disable.

With this Custom Scripts its not the best solution so there is a need for bringing this to good and working level.

Please @fichtner @AdSchellevis  and other look at this struggle to bring a good solution. 

br
sash

**Comment by fmeppo** [comment]
I've been running with scripts like this on my routers for a couple weeks now.  With minor fixes this provides reasonably robust backup, but failover takes ~30 sec (as the backup router must re-enable the interface, re-acquire the DHCP lease, etc.)  There's usually a noticeable hiccup when failover occurs for WAN traffic - but there isn't any for my local networks as these are all covered by CARP.

I'm wondering if a more robust solution would be to enable CARP on my WAN interface, claim a static IP equal to what DHCP hands out, then use a dhclient exit hook to reconfigure the virtual IP address on both primary and backup routers whenever the lease changes.  Obviously dhclient's default behaviors (directly changing the interface, updating resolv.conf, etc.) would need to be prevented for that interface.  Has anyone tried such an approach?

**Comment by LorenKeagle** [comment]
This is needed in a supported fashion for anyone doing multi-WAN failover with two DHCP WAN connections.  In my case, I have cable internet, and a cellular backup. 

I've tried several of the scripts available online, but they all either didn't work correctly in the first place, or stopped working when the underlying method signatures have changed.

Automatically enabling/disabling WAN interfaces based on CARP status seems like a no-brainer solution for those of us that do not have static IPs, but still want failover if a primary router goes down or needs to be reset for maintenance.

**Comment by oasis9** [comment]
I apologise for my lack of input, I would benefit from this as a feature but I struggle to understand the inner workings enough to confidently make these sorts of changes without the input of others who have had the time and knowledge to create the solutions we have so far. It's still going to be several months before I can even start thinking about switching to a provider with true static routing (I can't get anything other than a DHCP reservation right now, so can't do CARP). I'd appreciate if this could be incorporated into OPNsense itself, I've been using it to run my homelab and I can imagine many others have similar setups that would benefit from this functionality. I don't think this issue should be closed.

**Comment by 986ster** [comment]
Bumping this as a really useful feature for me too. My ISP only issues a single DHCP IPv4 address and I'd like to be able to fail over to a second node if there is anything up with my primary node or I need to take it offline for an update, etc., as well as making my backup cellular link available to both nodes. Putting another router (or two in my case) in front of the two HA nodes is not a good solution as it introduces another single point of failure, and the script solution seems to be perpetually on the edge of breaking with each new OPNsense version.

**Comment by lavacano** [comment]
I took a crack at it lmk how it works

https://gist.github.com/lavacano/a678e65d31df9bec344e572461ed3e10

with verbose 

<img width="720" height="916" alt="Image" src="https://github.com/user-attachments/assets/67af7b0e-3bf2-41a2-bd8b-6f4125402d64" />

**Comment by wiilover22** [comment]
> I took a crack at it lmk how it works
> 
> https://gist.github.com/lavacano/3e3c7f91ed34d5fad1cd58595d754525
> 


Your link no longer works - and I would also like to mention I would really like this feature baked in! I just spend an entire midnight fixing my network after learning how to setup HA - only to find out my modem gives me a single DHCP IPv4 address. 

Which of course means I can't apply updates to one router, while my other one takes over...

**Comment by lavacano** [comment]
> > I took a crack at it lmk how it works
> > https://gist.github.com/lavacano/3e3c7f91ed34d5fad1cd58595d754525
> 
> Your link no longer works - and I would also like to mention I would really like this feature baked in! I just spend an entire midnight fixing my network after learning how to setup HA - only to find out my modem gives me a single DHCP IPv4 address.
> 
> Which of course means I can't apply updates to one router, while my other one takes over...

If 4.79 doesn't work, there is an older 2.9 version on the gist. The 2.9 version is basically just a refactored version of the simple script that I'm sure was linked above. 


https://gist.github.com/lavacano/a678e65d31df9bec344e572461ed3e10

I'm still testing the following one.  But it should have a lot tighter control more logging more testing. The problem with dialing in problems is it adds complexity and it creates more problems, right? It's that adage I had 99 bugs and I fixed one and now I have 112 bugs. 

https://github.com/lavacano/OPNsense-HA-Failover

For any tunable in the comments of the 4.7.9 script or the 2.9 script I found that they don't work very well and what actually works is setting the tunable disable carp preempt to one on both routers then clicking the disable preempt check box in the ha config on the backup router.  

This allows the systems to come online and get healthy without trying to take vip. But once they are healthy ha loads and then the preempt might force a failback. 

Another idea to get rid of the bouncing that might happen is to configure the routers to not care which one is backup or master. The problem I've seen in this scenario if you have multiple VIPs on multiple vlans is sometimes the failover can happen on one VLAN but not all of them.
